### PR TITLE
feat: improve error location

### PR DESCRIPTION
**Pull request by fisker**
## PR Checklist

- [ ] Addresses an existing open issue: fixes #000
- [ ] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [ ] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

### CODE REVIEW

File: packages/typescript-estree/src/node-utils.ts [hunk 1]
```diff
@@ -611,21 +635,39 @@ export class TSError extends Error {
...
+    return this.location.start.offset;
+  }
+
+  get lineNumber(): number {
+    return this.location.start.line;
+  }
+
+  get column(): number {
+    return this.location.start.column;
+  }
```
Comment (fisker) [inline packages/typescript-estree/src/node-utils.ts:649]: Should those kept?

File: packages/typescript-estree/src/convert.ts [hunk 2]
```diff
@@ -2981,13 +2985,17 @@ export class Converter {
...
   }
 
-  #throwErrorIfDeprecatedPropertyExists<Node extends ts.Node>(
-    node: Node,
-    property: unknown,
-    message: string,
-  ): void {
-    if (property) {
-      throw createError(this.ast, node.pos, message);
+  #throwError(node: ts.Node | number, message: string): void {
```
Comment (armano2) [inline packages/typescript-estree/src/convert.ts:2988]:
```suggestion
  #throwError(
    node: ts.Node | number,
    message: string,
  ): asserts node is never {
```

with that you cna remove all type casts that you added eg,

```diff
- property: this.convertJSXIdentifier(node.name),
+ property: this.convertJSXIdentifier(node.name as ts.Identifier),
```

File: packages/typescript-estree/src/node-utils.ts (hunk 1 above)
Comment (fisker) [reply packages/typescript-estree/src/node-utils.ts:649]: Is this the answer? https://github.com/eslint/eslint/blob/b09a512107249a4eb19ef5a37b0bd672266eafdb/lib/linter/linter.js#LL853C32-L853C32

Comment (armano2) [reply packages/typescript-estree/src/node-utils.ts:649]: that's correct, we have to keep `lineNumber` and `column` as they are used in eslint
as for `index` if i remember correctly, this prop was needed in older version of eslint

File: packages/typescript-estree/src/convert.ts (hunk 2 above)
Comment (fisker) [reply packages/typescript-estree/src/convert.ts]: Thanks!

**Review by armano2 — APPROVED** [review]

File: packages/typescript-estree/src/convert.ts [hunk 3]
```diff
@@ -2981,13 +2985,17 @@ export class Converter {
...
   }
 
-  #throwErrorIfDeprecatedPropertyExists<Node extends ts.Node>(
-    node: Node,
-    property: unknown,
-    message: string,
-  ): void {
-    if (property) {
-      throw createError(this.ast, node.pos, message);
+  #throwError(node: ts.Node | number, message: string): asserts node is never {
```
Comment (armano2) [inline packages/typescript-estree/src/convert.ts:2988]:
```suggestion
  #throwError(node: ts.ReadonlyTextRange | number, message: string): asserts node is never {
```

and 

```
      start = node.start;
      end = node.end;
```

File: packages/typescript-estree/src/convert.ts [hunk 4]
```diff
@@ -3033,8 +3033,8 @@ export class Converter {
   #checkIllegalDecorators(node: ts.Node): void {
     if (nodeHasIllegalDecorators(node)) {
       this.#throwUnlessAllowInvalidAST(
-        // eslint-disable-next-line deprecation/deprecation
-        node.decorators?.[0] ?? node,
+        // @ts-expect-error -- this is safe as it's guarded
+        node.illegalDecorators?.[0] ?? node,
```
Comment (armano2) [inline packages/typescript-estree/src/convert.ts]: good catch, i'v tried to fix all conflicts and there was a lot, but it seem i missed this one

**Pull request merged**